frontend/src/pages/RunList.tsx

```diff
@@ -331,9 +331,12 @@ class RunList extends React.PureComponent<RunListProps, RunListState> {
 
     if (Array.isArray(this.props.runIdListMask)) {
       displayRuns = this.props.runIdListMask.map(id => ({ run: { id } }));
-      // listRuns doesn't currently support batching by IDs, so in this case we retrieve each run
-      // individually.
-      await this._getAndSetRuns(displayRuns);
+      const filter = JSON.parse(
+        decodeURIComponent(request.filter || '{"predicates": []}'),
+      ) as ApiFilter;
+      // listRuns doesn't currently support batching by IDs, so in this case we retrieve and filter
+      // each run individually.
+      await this._getAndSetRuns(displayRuns, filter);
```

Comment:
displayRuns = this.props.runIdListMask.map(id => ({ run: { id } }));
      await this._getAndSetRuns(displayRuns);

      const filter = JSON.parse(
        decodeURIComponent(request.filter || '{"predicates": []}'),
      ) as ApiFilter;
      const predicates = filter.predicates?.filter(p => p.key === 'name' && p.op === PredicateOp.ISSUBSTRING);
      const substrings = predicates?.map(p => p.string_value?.toLowerCase() || '');
      displayRuns = displayRuns.filter(runDetail => {
        for (let sub in substrings) {
          if (!(runDetail?.run?.name?.toLowerCase().includes(sub))) {
            return false;
          }
        }
        return true;
      });

How do you think about this logic? To explain:
1. Fetch all run details first.
2. Convert filter predicates to a list of `substrings`.
3. For each run's name, check whether it contains substrings. If not, remove this run from `displayRuns` list.

Reply:
That makes sense; I didn't realize that there could be multiple predicates checking substrings against the same value, this solution works much better in that case. I added another test to check this behavior.

I updated the `for` loop line to `for (const sub of substrings) {` since the `sub` value is never reassigned and the `for...of` loop is used for arrays.

Reply:
Make sense on the use of `const`! You can find definition of filter in `filter.proto`. And even more, this file describes advanced usage like IN: https://github.com/kubeflow/pipelines/blob/master/backend/api/filter.proto#L106-L117. But since frontend is not creating IN filter, we simplify the logic by only considering `ISSUBSTRING` predicateOp